### PR TITLE
ESM-2 Accelerate Recipes

### PR DESCRIPTION
**Pull request by pstjohn**
Initial work on adding an accelerate recipe for ESM-2. This will let us more easily compare TE to FA2 performance with accelerate and their built-in distributed settings (FSDP2, etc.)

I noticed a bug with the ESM-2 model implementation that we'll need to upstream to the hf model repo

## Summary by CodeRabbit

- New Features
  - Added an ESM-2 Accelerate recipe with Hydra configs, dataset/metrics utilities, callback to stop after N steps, and SLURM script for multi-node runs.
  - Provided FSDP-ready accelerate configs (single-process and single-node variants).

- Bug Fixes
  - Corrected parameter name to pretrained_model and improved distributed teardown safety.
  - Improved rotary embedding handling for stability across devices and sequence lengths.

- Documentation
  - Added READMEs for the new recipe and Hydra configs with run examples.

- Tests
  - Introduced end-to-end and multi-GPU launch tests with improved logging and timeouts.

- Chores
  - Updated dependencies, added lint/config files, dockerignore, and adjusted default accelerate settings.

### CODE REVIEW

File: models/esm2/src/esm/modeling_esm_te.py [hunk 1]
```diff
@@ -157,14 +157,20 @@ def forward(
         """
         all_hidden_states = () if output_hidden_states else None
 
+        if self.te_rope_emb is not None:
+            te_rope_emb = self.te_rope_emb.to(hidden_states.device, non_blocking=True)
+            te_rope_emb = te_rope_emb[: hidden_states.shape[1]]
+        else:
+            te_rope_emb = None
```
Comment (jomitchellnv) [inline models/esm2/src/esm/modeling_esm_te.py:177]: is none acceptable?

File: recipes/esm2_accelerate/hydra_config/defaults.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  per_device_eval_batch_size: ???
+  optim: "adamw_torch_fused"
+  learning_rate: 4e-4
+  weight_decay: 0.01
+  adam_beta1: 0.9
+  adam_beta2: 0.98
+  max_grad_norm: 1.0
+  adam_epsilon: 1e-8
+  num_train_epochs: 1
+  lr_scheduler_type: "linear"
```
Comment (jomitchellnv) [inline recipes/esm2_accelerate/hydra_config/defaults.yaml:20]: ESM2 was trained using the WarmupAnnealDecay LR scheduler. Does this one have decay on it?

File: recipes/esm2_accelerate/test_train.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+_fp8_available, _fp8_reason = check_fp8_support()
+
+requires_multi_gpu = pytest.mark.skipif(
+    not torch.cuda.is_available() or torch.cuda.device_count() < 2,
+    reason="Test requires at least 2 GPUs",
+)
+
+
+@pytest.fixture(scope="session")
```
Comment (jomitchellnv) [inline recipes/esm2_accelerate/test_train.py]: Ah curious what "scope" does here

File: recipes/esm2_accelerate/test_train.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+    assert len(model_files) > 0, f"No model files found in {final_checkpoint}"
+    assert len(config_files) > 0, f"No config files found in {final_checkpoint}"
+
+    # Check that training metrics were saved
+    train_metrics_file = output_dir / "train_results.json"
+    assert train_metrics_file.exists(), f"Training metrics file {train_metrics_file} does not exist"
+
+
+@pytest.mark.parametrize(
```
Comment (jomitchellnv) [inline recipes/esm2_accelerate/test_train.py:125]: I dont like parametrize and will always prefer having more tests

File: recipes/esm2_accelerate/test_train.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        print(f"STDERR:\n{result.stderr}")
+        pytest.fail(f"Command:\n{' '.join(cmd)}\nfailed with exit code {result.returncode}")
+
+
+@requires_multi_gpu
+@pytest.mark.parametrize(
+    "accelerate_config,model_tag",
+    [
+        ("default.yaml", "nvidia/esm2_t6_8M_UR50D"),
+        # Currently failing for some reason with device types not matching, oddly a local modeling_esm_te import seems
```
Comment (jomitchellnv) [inline recipes/esm2_accelerate/test_train.py]: ticket to track?

File: recipes/esm2_accelerate/test_train.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        cmd,
+        check=False,
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        timeout=240,
+    )
+
+    if result.returncode != 0:
+        print(f"STDOUT:\n{result.stdout}")
```
Comment (jomitchellnv) [inline recipes/esm2_accelerate/test_train.py:231]: does logging work instead of print?

**Review by jomitchellnv — APPROVED** [review]
LGTM a few nits

File: models/esm2/src/esm/modeling_esm_te.py (hunk 1 above)
Comment (pstjohn) [reply models/esm2/src/esm/modeling_esm_te.py:177]: none implies you're doing positional embeddings, no rope. Idk if it works honestly, i'm not sure we test this code path.

File: recipes/esm2_accelerate/hydra_config/defaults.yaml (hunk 2 above)
Comment (pstjohn) [reply recipes/esm2_accelerate/hydra_config/defaults.yaml:20]: yeah linear implies the linear decay, and there's a separate warmup parameter. this doesn't do the minimum lr bit that ESM-2 has

File: recipes/esm2_accelerate/test_train.py (hunk 3 above)
Comment (pstjohn) [reply recipes/esm2_accelerate/test_train.py]: hmm maybe i don't want this anymore 😄 I used to have separate tests for start and resume

File: recipes/esm2_accelerate/test_train.py (hunk 4 above)
Comment (pstjohn) [reply recipes/esm2_accelerate/test_train.py:125]: yeah, i'm OK with it when the parameter values are very clear and you don't need any test branching, but i know that AV slide deck discouraged them

File: recipes/esm2_accelerate/test_train.py (hunk 6 above)
Comment (pstjohn) [reply recipes/esm2_accelerate/test_train.py:231]: maybe but in the test we want to just make sure this hits stdout

File: recipes/esm2_accelerate/test_train.py (hunk 3 above)
Comment (pstjohn) [reply recipes/esm2_accelerate/test_train.py]: ok, removed

File: recipes/esm2_accelerate/test_train.py (hunk 5 above)
Comment (pstjohn) [reply recipes/esm2_accelerate/test_train.py]: thanks, added!

File: recipes/esm2_accelerate/train.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    trainer = Trainer(
+        model=model,
+        args=training_args,
+        train_dataset=train_dataset,
+        eval_dataset=eval_dataset,
+        compute_metrics=compute_metrics,
+        data_collator=data_collator,
+        callbacks=[StopAfterNStepsCallback(args.stop_after_n_steps)],
+    )
```
[the remark this answers is not in the document]
Comment (pstjohn) [reply recipes/esm2_accelerate/train.py:58]: this is a typing error in hf transformers.

**Review by jstjohn — APPROVED** [review]

**Comment by pstjohn** [comment]
@coderabbitai resolve

**Comment by pstjohn** [comment]
@coderabbitai resolve
